File: src/main/java/com/kevinjanvier/service/PetService.java
package com.kevinjanvier.service;

import com.kevinjanvier.entity.Pet;
import com.kevinjanvier.repository.PetRepository;

import javax.inject.Singleton;
import java.util.List;

@Singleton
public class PetService {

    private final PetRepository petRepository;

    public PetService(PetRepository petRepository) {
        this.petRepository = petRepository;
    }

    public Pet savePet(Pet pet) {
        return petRepository.save(pet);
    }

    public List<Pet> getAllPets() {
        return petRepository.findAll();
    }

    public Pet updatePet(Pet pet) {
        return petRepository.update(pet);
    }

    public void deletePet(Long id) {
        petRepository.deleteById(id);
    }
}

File: src/main/java/com/kevinjanvier/service/PetService.kt
package com.kevinjanvier.service

import com.kevinjanvier.entity.PetModel
import com.kevinjanvier.exception.NotFoundException
import javax.inject.Singleton
import com.kevinjanvier.repository.PetRepository
import org.bson.BsonValue

@Singleton
class PetService(private val petRepository: PetRepository) {

    fun createPet(request: PetModel): BsonValue? {
        val insertedPet = petRepository.create(
                request
        )
        return insertedPet!!.insertedId
    }

    fun findAll(): List<PetModel> {
        return petRepository.findAll()
    }

    fun findById(id: String): PetModel {
        return petRepository.findById(id)
                ?: throw NotFoundException("pet id doesn't exist")
    }

    fun updatePet(id: String, request: PetModel): PetModel {
        val updateResult = petRepository.update(id, request)

        if (updateResult!!.modifiedCount == 0L)
            throw throw RuntimeException("Pet id with id $id was not updated")

        val updatedPet = findById(id)
        updateResultForPet(updatedPet)
        return updatedPet
    }

    fun deleteById(id: String) {
        val deleteResult = petRepository.deleteById(id)
        if (deleteResult!!.deletedCount == 0L)
            throw throw RuntimeException("Pet id with id $id  was not deleted")

    }

    private fun updateResultForPet(updatedPet: PetModel) {
        petRepository.findAllByPetId(updatedPet.id!!.toHexString())
                .map {
                    petRepository.update(
                            it.id!!.toHexString(),
                            it.apply { it.image }
                    )
                }
    }
}
